### PR TITLE
Add DeepSeek-V4  DCP decode support

### PR DESCRIPTION
**Pull request by baonudesifeizhai**
## Summary

This PR adds decode context parallel support for DeepSeek-V4 Flash.

Main changes:
- Add DCP-aware FlashMLA decode output/LSE merge.
- Add DCP-aware compressor stats merge and compressed KV cache write.
- Add DCP local KV/index mapping for DeepSeek-V4 sparse decode paths.
- Keep existing PCP/prefill behavior unchanged.

## Test Plan
```
export MODEL=/root/model/DeepSeek-V4-Flash
export SERVED=deepseek-v4-flash

CUDA_VISIBLE_DEVICES=2,3 \
vllm serve "$MODEL" \
  --served-model-name "$SERVED" \
  --host 0.0.0.0 \
  --port 8000 \
  --tensor-parallel-size 2 \
  --decode-context-parallel-size 2 \
  --enable-expert-parallel \
  --moe-backend deep_gemm_mega_moe \
  --all2all-backend deepep_high_throughput \
  --kv-cache-dtype fp8_ds_mla \
  --max-model-len 131072 \
  --max-num-batched-tokens 16384 \
  --max-num-seqs 16 \
  --gpu-memory-utilization 0.95
```
```
 OPENAI_API_KEY=EMPTY LMEVAL_LOG_LEVEL=INFO \
.venv/bin/python -m lm_eval run \
  --model local-completions \
  --model_args 'model=deepseek-v4-flash,base_url=http://127.0.0.1:8000/v1/completions,tokenizer_backend=none,tokenized_requests=False,max_gen_toks=256,num_concurrent=16,timeout=600' \
  --tasks gsm8k \
  --num_fewshot 5 \
  --batch_size 1 \
  --log_samples \
  --output_path /tmp/dsv4_flash_dcp2_gsm8k_lmeval_full
```
```
2026-06-04:19:27:27 INFO     [loggers.evaluation_tracker:119] Saving per-task samples to /tmp/dsv4_flash_dcp2_gsm8k_lmeval_full/deepseek-v4-flash/*.jsonl
local-completions ({'model': 'deepseek-v4-flash', 'base_url': 'http://127.0.0.1:8000/v1/completions', 'tokenizer_backend': 'none', 'tokenized_requests': False, 'max_gen_toks': 256, 'num_concurrent': 16, 'timeout': 600}), gen_kwargs: ({}), limit: None, num_fewshot: 5, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.95|±  | 0.006|
|     |       |strict-match    |     5|exact_match|↑  | 0.95|±  | 0.006|
```
```
  export MODEL=/root/model/DeepSeek-V4-Pro
export SERVED=deepseek-v4-pro

CUDA_VISIBLE_DEVICES=4,5,6,7 \
vllm serve "$MODEL" \
  --served-model-name "$SERVED" \
  --host 0.0.0.0 \
  --port 8000 \
  --tensor-parallel-size 4 \
  --decode-context-parallel-size 2 \
  --enable-expert-parallel \
  --moe-backend deep_gemm_mega_moe \
  --all2all-backend deepep_high_throughput \
  --kv-cache-dtype fp8_ds_mla \
  --max-model-len 131072 \
  --max-num-batched-tokens 16384 \
  --max-num-seqs 16 \
  --gpu-memory-utilization 0.95
```
```
OPENAI_API_KEY=EMPTY LMEVAL_LOG_LEVEL=INFO \
.venv/bin/python -m lm_eval run \
  --model local-completions \
  --model_args 'model=deepseek-v4-pro,base_url=http://127.0.0.1:8000/v1/completions,tokenizer_backend=none,tokenized_requests=False,max_gen_toks=256,num_concurrent=16,timeout=600' \
  --tasks gsm8k \
  --num_fewshot 5 \
  --batch_size 1 \
  --log_samples \
  --output_path /tmp/dsv4_pro_tp4_dcp2_gsm8k_lmeval_full
```
```
 num_fewshot: 5, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9522|±  |0.0059|
|     |       |strict-match    |     5|exact_match|↑  |0.9530|±  |0.0058|
```
## Test Result
 
DecodeBench, DeepSeek-V4 Flash, TP2 vs TP2+DCP2, C=128, OSL=4096.

| Context | Mode | Output TPS | TPOT | TTFT | Peak KV | Peak Running |
|---:|---|---:|---:|---:|---:|---:|
| 0 | TP2 | 5977.14 tok/s | 21.38 ms | 116.71 ms | 3.06% | 128 |
| 0 | TP2+DCP2 | 4058.95 tok/s | 31.51 ms | 128.10 ms | 1.67% | 128 |
| 16K | TP2 | 5899.66 tok/s | 20.44 ms | 3917.57 ms | 7.68% | 128 |
| 16K | TP2+DCP2 | 4132.16 tok/s | 29.83 ms | 3560.10 ms | 4.09% | 128 |
| 64K | TP2 | 4437.97 tok/s | 21.64 ms | 19040.81 ms | 21.43% | 128 |
| 64K | TP2+DCP2 | 3349.44 tok/s | 30.61 ms | 20117.72 ms | 11.10% | 128 |
| 128K | TP2 | 3328.18 tok/s | 22.10 ms | 43088.40 ms | 39.77% | 127 |
| 128K | TP2+DCP2 | 2715.27 tok/s | 30.70 ms | 43700.32 ms | 20.45% | 128 |
| 256K | TP2 | 2172.72 tok/s | 23.20 ms | 91663 ms | 58.42% | 96 |
| 256K | TP2+DCP2 | 1935.67 tok/s | 30.12 ms | 93623 ms | 37.83% | 122 |
| 512K | TP2 | 1262.54 tok/s | 21.62 ms | 202190 ms | 76.59% | 64 |
| 512K | TP2+DCP2 | 1189.72 tok/s | 27.61 ms | 203657 ms | 39.02% | 64 |
| 768K | TP2 | 864.83 tok/s | 21.56 ms | 317780 ms | 57.26% | 32 |
| 768K | TP2+DCP2 | 841.13 tok/s | 25.97 ms | 314920 ms | 29.12% | 32 |
 

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/models/deepseek_v4/nvidia/flashmla.py [hunk 1]
```diff
@@ -259,6 +329,7 @@ def _forward_decode(
         # We treat queries in the same seq as different queries
         # and later we only attend by generated indices.
         # q arrives pre-padded to layer.padded_heads by the outer wrapper.
+        q, dcp_group, use_dcp = _maybe_gather_dcp_q(q)
```
Comment (GirasoleY) [inline vllm/models/deepseek_v4/nvidia/flashmla.py]: I think the q here is padded to 64 because flashmla only supports 64/128 heads, and this probably will fail with DCP > 2 case. We can gather the unpadded q and pad after gather (as needed) here.

File: vllm/model_executor/layers/sparse_attn_indexer.py
Comment (GirasoleY) [inline vllm/model_executor/layers/sparse_attn_indexer.py:613]: For my understanding, after this local topK, should there be a global topK and let each rank pick their local tokens that belong to the global topK? 

(basically an equivalend of _topk_per_row_prefill_dcp in the decode path, maybe I missed the code path.)

File: vllm/v1/attention/backends/mla/flashmla_sparse.py [hunk 2]
```diff
@@ -1109,28 +1131,61 @@ def _build_c128a_topk_metadata_kernel(
 
     if is_decode:
         # --- Decode: block-table lookup → global slot ids + count ---
-        is_valid_token = tl.load(slot_mapping_ptr + token_idx) >= 0
+        if DCP_WORLD_SIZE > 1:
+            is_valid_token = True
```
Comment (GirasoleY) [inline vllm/v1/attention/backends/mla/flashmla_sparse.py]: I thought this can be padded token and not a real token. Could you double check if this is correct?

**Review by GirasoleY — COMMENTED** [review]
Did a very rough scan.

Like the kv cache design :) Left a few questions inline for clarification.

**Comment by baonudesifeizhai** [comment]
Context | Mode | Output TPS | TPOT | TTFT | Peak GPU KV
-- | -- | -- | -- | -- | --
0 | TP4 | 6536.76 tok/s | 19.55 ms | 124.87 ms | 2.6%
0 | TP4+DCP4 | 4250.01 tok/s | 30.06 ms | 237.66 ms | 0.9%
16K | TP4 | 4063.01 tok/s | 25.26 ms | 25.14 s | 12.1%
16K | TP4+DCP4 | 2465.17 tok/s | 40.37 ms | 46.56 s | 3.4%
64K | TP4 | 1826.79 tok/s | 43.92 ms | 105.10 s | 24.4%
64K | TP4+DCP4 | 1339.70 tok/s | 61.95 ms | 135.25 s | 6.5%
128K | TP4 | 992.61 tok/s | 72.27 ms | 228.03 s | 41.1%
128K | TP4+DCP4 | 1034.45 tok/s | 74.14 ms | 199.79 s | 10.7%
256K | TP4 | 468.74 tok/s | 141.73 ms | 529.01 s | 73.9%
256K | TP4+DCP4 | 559.39 tok/s | 123.77 ms | 423.59 s | 19.2%

for tp 4+dcp4 
image

**Comment by baonudesifeizhai** [comment]
CUDA_VISIBLE_DEVICES=4,5,6,7 \
VLLM_TEST_DSV4_MODEL=/root/model/DeepSeek-V4-Flash \
.venv/bin/python -m pytest \
  tests/distributed/test_context_parallel.py \
  -k deepseek_v4_dcp_end_to_end -s  --passed

File: vllm/v1/attention/ops/cp_utils.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+import torch
+
+from vllm.triton_utils import tl, triton
+
+if TYPE_CHECKING:
+    from vllm.config import VllmConfig
+
+
+@dataclass(frozen=True, slots=True)
+class ContextParallelLayout:
```
Comment (GirasoleY) [inline vllm/v1/attention/ops/cp_utils.py]: 👍🏻 Thanks for deduping the *dcp*pcp scaling formula that are currently scattered in the codebase + adding the slot mapping into one util class.

Minor nits:
* we can move the cp scaling concept into `KVCacheSpec` and allow some specific spec to not be scaled with CP (i.e. linear layer). 
* the rest changes can go in `vllm/v1/worker/cp_utils.py` Probably do not create a new file.

File: vllm/v1/context_parallel/collectives.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+) -> torch.Tensor | tuple[torch.Tensor, torch.Tensor]:
+    """Reduce partial attention output/LSE across DCP ranks."""
+    return dcp_a2a_lse_reduce(
+        local_out,
+        local_lse,
+        group,
+        ctx=ctx,
+        return_lse=return_lse,
+        is_lse_base_on_e=is_lse_base_on_e,
+    )
```
Comment (GirasoleY) [inline vllm/v1/context_parallel/collectives.py]: just keep existing `dcp_a2a_lse_reduce`?

File: vllm/v1/context_parallel/collectives.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+
+class AttentionOutputReducer:
```
Comment (GirasoleY) [inline vllm/v1/context_parallel/collectives.py]: Reading the methods I feel we don't need a class here, just plan util functions will do.

File: vllm/v1/context_parallel/collectives.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (GirasoleY) [inline vllm/v1/context_parallel/collectives.py]: I feel these are not really collectives but attention/compressor ops,  should go into something like `ops/dcp.py` (maybe just rename the current `dcp_alltoall.py`)

File: vllm/models/deepseek_v4/nvidia/flashmla.py
Comment (GirasoleY) [inline vllm/models/deepseek_v4/nvidia/flashmla.py]: The topK here is a various length, but since flashmla reuses scheduler metadata, I think the kernel expect the `topk_indices` to be padded to a fixed length that aligns with what is in metadata.

Otherwise we might end up with Nan/data corruption I think.

File: vllm/models/deepseek_v4/attention.py [hunk 7]
```diff
@@ -614,6 +616,7 @@ def get_kv_cache_spec(self, vllm_config: VllmConfig) -> KVCacheSpec | None:
             cache_dtype_str=self.kv_cache_dtype,
             alignment=576 if is_flashmla else None,  # FlashMLA needs 576B
             model_version="deepseek_v4",
+            supports_context_parallel=True,
```
Comment (GirasoleY) [inline vllm/models/deepseek_v4/attention.py]: Could you guard this with flashmla backend only?

**Review by GirasoleY — COMMENTED** [review]
Thanks for working on the DSv4 DCP prototype! 

A few high level comments:

- If I read correctly the current implementation does gatherQ on both prefill & decode path and only support A2A backend. Maybe just assert this behavior in the configuration?

- For indexer/compressor logic, the current implementation looks right :) (We can follow up with optimizations but I think this is a good foundation to land in)

- For cache layout change, this might conflict with https://github.com/vllm-project/vllm/pull/40996, since 40996 is a more contained change we will likely merge that first. Please keep an eye and rebase after. 
 
- Like the idea of unifying cp utils, but I think we don't need to create a top level folder for it: 
  - sharding fact -> KVCacheSpec (cp_world_size)
  - layout + triton index helpers -> cp_utils.py
  - attention-output merge (apply_attn_sink, dcp_a2a_lse_reduce) + compressor stats merge (dcp_softmax_reduce) -> `deepseek_v4/common/ops/dcp.py`

**Comment by GirasoleY** [comment]
Also once the merge conflict is resolved I'll add the ci tag.

**Comment by baonudesifeizhai** [comment]
I will update benchmark when i wake up

**Comment by baonudesifeizhai** [comment]
TP4+DCP4, C=128, OSL=4096

ctx=0      out_tps=4234.79  TPOT=30.20ms   TTFT=0.13s    peak_kv=1.0%
ctx=16K    out_tps=2371.49  TPOT=41.41ms   TTFT=50.73s   peak_kv=3.3%
ctx=64K    out_tps=1259.91  TPOT=65.09ms   TTFT=147.20s  peak_kv=6.5%
ctx=128K   out_tps=994.63   TPOT=78.17ms   TTFT=203.77s  peak_kv=10.5%
ctx=262K   out_tps=538.02   TPOT=131.71ms  TTFT=428.54s  peak_kv=19.0%


local-completions ({'model': 'deepseek-v4-flash', 'base_url': 'http://127.0.0.1:8000/v1/completions', 'tokenizer_backend': 'none', 'tokenized_requests': False, 'max_gen_toks': 256, 'num_concurrent': 64, 'timeout': 1800}), gen_kwargs: ({}), limit: None, num_fewshot: 5, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9462|±  |0.0062|
|     |       |strict-match    |     5|exact_match|↑  |0.9469|±  |0.0062|
